src/interpreter/mod.rs

```diff
@@ -1169,58 +1222,89 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
-                self.write_value_to_ptr(src_val, ptr, dest_ty)?;
+                self.write_value_to_ptr(src_val, ptr, dest_ty)
             }
 
-            // The cases here can be a bit subtle. Read carefully!
             Lvalue::Local { frame, local } => {
-                let dest_val = self.stack[frame].get_local(local);
-
-                if let Some(Value::ByRef(dest_ptr)) = dest_val {
-                    // If the local value is already `ByRef` (that is, backed by an `Allocation`),
-                    // then we must write the new value into this allocation, because there may be
-                    // other pointers into the allocation. These other pointers are logically
-                    // pointers into the local variable, and must be able to observe the change.
-                    //
-                    // Thus, it would be an error to replace the `ByRef` with a `ByVal`, unless we
-                    // knew for certain that there were no outstanding pointers to this local.
-                    self.write_value_to_ptr(src_val, dest_ptr, dest_ty)?;
-
-                } else if let Value::ByRef(src_ptr) = src_val {
-                    // If the local value is not `ByRef`, then we know there are no pointers to it
-                    // and we can simply overwrite the `Value` in the locals array directly.
-                    //
-                    // In this specific case, where the source value is `ByRef`, we must duplicate
-                    // the allocation, because this is a by-value operation. It would be incorrect
-                    // if they referred to the same allocation, since then a change to one would
-                    // implicitly change the other.
-                    //
-                    // TODO(solson): It would be valid to attempt reading a primitive value out of
-                    // the source and writing that into the destination without making an
-                    // allocation. This would be a pure optimization.
-                    let dest_ptr = self.alloc_ptr(dest_ty)?;
-                    self.copy(src_ptr, dest_ptr, dest_ty)?;
-                    self.stack[frame].set_local(local, Value::ByRef(dest_ptr));
-
-                } else {
-                    // Finally, we have the simple case where neither source nor destination are
-                    // `ByRef`. We may simply copy the source value over the the destintion local.
-                    self.stack[frame].set_local(local, src_val);
-                }
+                let dest = self.stack[frame].get_local(local);
+                self.write_value_to_dest(
+                    src_val,
+                    |this, val| this.stack[frame].set_local(local, val),
+                    dest,
+                    dest_ty,
+                )
             }
         }
+    }
+
+    // The cases here can be a bit subtle. Read carefully!
+    fn write_value_to_dest<F: FnOnce(&mut Self, Value)>(
```

Comment:
The comments in this function reference locals, but you made it more general than locals.

I also don't like the name. `write_value_to_dest` just sounds like a synonym of `write_value`.

To be really explicit, it could be `write_value_possibly_by_val` or something along those lines.